airflow/utils/db.py

```diff
@@ -626,27 +629,102 @@ def check_migrations(timeout):
         context = MigrationContext.configure(connection)
-        ticker = 0
-        while True:
-            source_heads = set(script_.get_heads())
-            db_heads = set(context.get_current_heads())
-            if source_heads == db_heads:
-                break
-            if ticker >= timeout:
-                raise TimeoutError(
-                    f"There are still unapplied migrations after {ticker} seconds. "
-                    f"Migration Head(s) in DB: {db_heads} | Migration Head(s) in Source Code: {source_heads}"
-                )
-            ticker += 1
-            time.sleep(1)
-            log.info('Waiting for migrations... %s second(s)', ticker)
+        return set(context.get_current_heads())
+
+
+def check_and_run_migrations():
+    """Check and run migrations if necessary. Only use in a tty"""
+    from rich import print as rich_print
+
+    source_heads = get_source_heads()
+    db_heads = get_db_heads()
+    if len(db_heads) < 1:
+        if sys.stdout.isatty():
+            print()
+            question = (
+                "Please confirm database initialization " "(or wait 4 seconds to skip it). Are you sure?"
+            )
+            try:
+                ans = helpers.prompt_with_timeout(question, timeout=4)
+                if ans:
+                    # initialize database
+                    print("DB: " + repr(settings.engine.url))
+                    initdb()
+                    print("DB Initialization done")
+            except AirflowException:
+                pass
+        else:
+            rich_print(
+                "[red][bold]ERROR:[/bold] You need to initialize the database. "
+                "Run `airflow db init` to initialize the database"
+            )
+            sys.exit(1)
+    elif source_heads != db_heads:
+        if sys.stdout.isatty():
+            print()
+            question = "Please confirm database upgrade" " (or wait 4 seconds to skip it). Are you sure?"
+            try:
+                ans = helpers.prompt_with_timeout(question, timeout=4)
+                if ans:
+                    # upgrade database
+                    try:
+                        print("DB: " + repr(settings.engine.url))
+                        upgradedb()
+                        print("DB Upgrade done")
+                    except Exception as e:
+                        print(e)
+                        rich_print(
+                            "[red][bold]ERROR:[/bold] You still have unapplied migrations. "
+                            "You may need to reset the database by running `airflow db reset`"
+                        )
+                        sys.exit(1)
+            except AirflowException:
+                pass
+        else:
+            rich_print(
+                "[red][bold]ERROR:[/bold] You need to upgrade the database. "
+                "Run `airflow db upgrade` to upgrade"
+            )
+            sys.exit(1)
```

Comment:
We can reduce the duplication in these blocks a lot by doing something like this

```suggestion
def check_and_run_migrations():
    """Check and run migrations if necessary. Only use in a tty"""
    from rich import print as rich_print

    source_heads = get_source_heads()
    db_heads = get_db_heads()
    if len(db_heads) < 1:
        db_command = initdb
        db_name = "init"
    else:
        db_command = upgradedb
        db_name = "upgrade"
        
        if sys.stdout.isatty():
            print()
            question = (
                "Please confirm database initialization " "(or wait 4 seconds to skip it). Are you sure?"
            )
            try:
                ans = helpers.prompt_with_timeout(question, timeout=4)
                if ans:
                    # initialize database
                    print("DB: " + repr(settings.engine.url))
                    initdb()
                    print("DB Initialization done")
            except AirflowException:
                pass
        else:
            rich_print(
                "[red][bold]ERROR:[/bold] You need to initialize the database. "
                "Run `airflow db init` to initialize the database"
            )
            sys.exit(1)
    elif source_heads != db_heads:
        if sys.stdout.isatty():
            print()
            question = "Please confirm database upgrade" " (or wait 4 seconds to skip it). Are you sure?"
            try:
                ans = helpers.prompt_with_timeout(question, timeout=4)
                if ans:
                    # upgrade database
                    try:
                        print("DB: " + repr(settings.engine.url))
                        upgradedb()
                        print("DB Upgrade done")
                    except Exception as e:
                        print(e)
                        rich_print(
                            "[red][bold]ERROR:[/bold] You still have unapplied migrations. "
                            "You may need to reset the database by running `airflow db reset`"
                        )
                        sys.exit(1)
            except AirflowException:
                pass
        else:
            rich_print(
                "[red][bold]ERROR:[/bold] You need to upgrade the database. "
                "Run `airflow db upgrade` to upgrade"
            )
            sys.exit(1)
```
```